### PR TITLE
✨ Improve MD patching with helper

### LINKED ISSUE #1908: Missing cluster label for machine deployment
**What steps did you take and what happened:**
I created a CAPA cluster using `make cluster-create` and did a `kubectl apply -f examples/_out/machinedeployment.yaml`.

I noticed that the machine deployment didn't have cluster label.
```
$ kubectl get machinedeployments --show-labels
NAME         PHASE       LABELS
test2-md-0   ScalingUp   nodepool=nodepool-0
```
However, it did have the correct owner reference and `spec.ClusterName`. MachineSets and Machines had the correct cluster label.

**What did you expect to happen:**
I expected the machine deployment to have the correct cluster label.

/kind bug

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
This PR cleans up the MachineDeployment controller to:
- Always patch the deployment and its status after finishing the reconciliation loop, in line with what we do today in other controllers.
- Adds a test to check for that the Cluster OwnerReference and Label is applied on the MachineDeployment.
- Change any "normal" patch to use the patch helper instead.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to https://github.com/kubernetes-sigs/cluster-api/issues/1690
Fixes #1908 

/assign @ncdc

### CODE REVIEW

File: controllers/machinedeployment_sync.go [hunk 1]
```diff
@@ -104,19 +103,22 @@ func (r *MachineDeploymentReconciler) getNewMachineSet(d *clusterv1.MachineDeplo
 	// latest revision.
 	if existingNewMS != nil {
 		msCopy := existingNewMS.DeepCopy()
-		patch := client.MergeFrom(msCopy.DeepCopy())
+		patchHelper, err := patch.NewHelper(msCopy, r.Client)
```
Comment (ncdc) [inline controllers/machinedeployment_sync.go]: Can we get rid of these "inner" patch helpers?

Comment (vincepri) [reply controllers/machinedeployment_sync.go]: I don't think we can safely do that today, unless we have a way to keep track of the old/new MachineSet the MachineDeployment is managing

**Comment by ncdc** [comment]
/lgtm

Let's keep an eye on CAPA CI after this merges

**Pull request merged**